ESQL: Add url_encode function

mouhc1ne opened the pull request:
- Implement `url_encode` function

mouhc1ne committed:
ESQL: Add url_encode function

mouhc1ne commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/UrlEncode.java:
```diff
@@ -0,0 +1,87 @@
+    @ConvertEvaluator()
+    static BytesRef process(final BytesRef val) {
+        String s = val.utf8ToString();
+        String encoded = URLEncoder.encode(s, StandardCharsets.UTF_8);
```
Two concerns I have about this snippet:

1. This actually over-encodes the input. For example `http://elastic.co` becomes `http%3A%2F%2Felastic.co`. I wonder if we have a better option internally.
2. Performance when passing strings around, as I see all other evaluators directly manipulating byte arrays.

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/UrlEncodeTests.java:
```diff
@@ -0,0 +1,72 @@
+        String path = randomFrom("", "/" + randomAlphanumericOfLength(5) + "/");
+        String query = randomFrom("", "?" + randomAlphaOfLength(5) + "=" + randomAlphanumericOfLength(5));
+        return String.format("%s%s%s%s", protocol, domain, path, query);
+    }
```
It'd be nice to return a random input url and a random encoded one. Like
```
   record RandomUrl(String plain, String encoded) {}

    private static RandomUrl generateRandomUrl() {

```

That way you can encode it by hand and not rely on the built in url-encode to sort of double prove it.

ivancea reviewed (commented):
LGTM!

mouhc1ne committed:
ESQL: Add url_encode function
- Rename the capability to exclude the decoding portion, since this PR is only concerned with encoding.
- Mark url_encode as SNAPSHOT to allow changes in encoding logic.
- Add tests that use random strings to test a wide range of characters, that run alongside the ones that use random URLs.
- Minor change in the sample csv test string to showcase a fictional URL with path and few query params.

mouhc1ne committed:
ESQL: Add url_encode function
- Remove url_encode from docs since it's a snapshot function
- Declare it along other snapshot functions in EsqlFunctionRegistry
- Annotate UrlEncode similarly to other snapshot functions
- Add url_encode capability to csv tests

mouhc1ne commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -1389,7 +1389,12 @@ public enum Cap {
+        /**
+         * URL encoding and decoding functions.
+         */
+        URL_DECODE_ENCODE;
```
Renamed to `URL_ENCODE`. Initially I was thinking of adding the decode function right into this PR after getting some feedback, but now I see one PR for each to keep things moving, so it makes sense to rename.

nik9000 reviewed (approved):
Looks good to me too.

mouhc1ne commented:
The failing release-test (https://buildkite.com/elastic/elasticsearch-pull-request/builds/88528/steps/canvas?jid=0198f0c9-4928-4cf3-9939-0a867d857f5f) is unrelated to these changes. Removing `test-release` label.
